## Fine-tuning the GPT Model using Lora fine-tuning technique

1. You can change model_name to a smaller one like EleutherAI/pythia-70m if needed or also you can replace model with ollama, llama 2, etc.,

2. Since we're running it from colab, we should use less weight models for practice

3. This example uses qLoRA (load_in_4bit=True) + LoRA.

In [1]:
!pip install transformers datasets peft accelerate bitsandbytes evaluate rouge_score trl -q

  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.7/60.7 MB 13.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.1/84.1 kB 4.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 697.4/697.4 kB 13.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 527.0/527.0 kB 11.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 47.6/47.6 MB 14.9 MB/s eta 0:00:00


# Summarization dataset to train the model

In [2]:
# ========================
# STEP 2: Load Dataset and Tokenizer
# ========================
from datasets import load_dataset
from transformers import AutoTokenizer

dataset = load_dataset("knkarthick/samsum")
tokenizer = AutoTokenizer.from_pretrained("tiiuae/falcon-rw-1b")

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:103: UserWarning: 
Error while fetching `HF_TOKEN` secret value from your vault: 'Requesting secret HF_TOKEN timed out. Secrets can only be fetched when running from the Colab UI.'.
You are not authenticated with the Hugging Face Hub in this notebook.
If the error persists, please let us know by opening an issue on GitHub (https://github.com/huggingface/huggingface_hub/issues/new).
  warnings.warn(


README.md: 0.00B [00:00, ?B/s]

train.csv: 0.00B [00:00, ?B/s]

validation.csv: 0.00B [00:00, ?B/s]

test.csv: 0.00B [00:00, ?B/s]

Generating train split:   0%|          | 0/14731 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/818 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/819 [00:00<?, ? examples/s]

config.json: 0.00B [00:00, ?B/s]

tokenizer_config.json:   0%|          | 0.00/234 [00:00<?, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/99.0 [00:00<?, ?B/s]

# Method to Evaluate the LLM models

In [3]:

# ========================
# STEP 3: Evaluate Base Model
# ========================

from transformers import AutoModelForCausalLM
import evaluate
import torch

rouge = evaluate.load("rouge")
base_model = AutoModelForCausalLM.from_pretrained("tiiuae/falcon-rw-1b").to("cuda")
base_model.eval()

def evaluate_model(model, dataset, tokenizer, n=20):
    preds, refs = [], []
    for example in dataset.select(range(n)):
        prompt = f"Summarize this dialogue:\n\n{example['dialogue']}\n\nSummary:"
        inputs = tokenizer(prompt, return_tensors="pt").to("cuda")
        with torch.no_grad():
            output = model.generate(**inputs, max_new_tokens=60)
        summary = tokenizer.decode(output[0], skip_special_tokens=True).split("Summary:")[-1]
        preds.append(summary.strip())
        refs.append(example["summary"])

        # --- SNIPPET TO SEE SAMPLES ---
        print(f"REFERENCE: {example['summary']}")
        print(f"PREDICTION: {summary}")

    return rouge.compute(predictions=preds, references=refs)

base_score = evaluate_model(base_model, dataset["test"], tokenizer)
print("\n" + "="*30)
print("ROUGE (Base Model):", base_score)

pytorch_model.bin:   0%|          | 0.00/2.62G [00:00<?, ?B/s]

Loading weights:   0%|          | 0/292 [00:00<?, ?it/s]

model.safetensors:   0%|          | 0.00/2.62G [00:00<?, ?B/s]

The tied weights mapping and config for this model specifies to tie transformer.word_embeddings.weight to lm_head.weight, but both are present in the checkpoints, so we will NOT tie them. You should update the config with `tie_word_embeddings=False` to silence this warning


generation_config.json:   0%|          | 0.00/115 [00:00<?, ?B/s]

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


REFERENCE: Hannah needs Betty's number but Amanda doesn't have it. She needs to contact Larry.
PREDICTION: 
Amanda and Hannah are best friends.
Notes:
- I'm sorry for the delay. I was sick for a few days.
- I'm sorry for the lack of updates. I was sick for a few days.
- I'm sorry for the lack of chapters. I


Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


REFERENCE: Eric and Rob are going to watch a stand-up on youtube.
PREDICTION: 
Rob: I'm so happy to see you again!
Eric: I'm so happy to see you too!
Rob: I'm so happy to see you again!
Eric: I'm so happy to see you again!
Rob: I'm so happy to see you again!


Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


REFERENCE: Lenny can't decide which trousers to buy. Bob advised Lenny on that topic. Lenny goes with Bob's advice to pick the trousers that are of best quality.
PREDICTION: 
Lenny: Babe, can you help me with something?
Bob: Which one should I pick?
Lenny: Which one should I pick?
Bob: Send me photos
Lenny:  <file_photo>
Lenny:  <file_photo>
L


Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


REFERENCE: Emma will be home soon and she will let Will know.
PREDICTION: 
Will and Emma are in a relationship.
Will: hey babe, what do you want for dinner tonight?
Emma: 
Will: what do you mean? everything ok?
Emma: not really, but it's ok, don't worry about cooking though, I'm


Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


REFERENCE: Jane is in Warsaw. Ollie and Jane has a party. Jane lost her calendar. They will get a lunch this week on Friday. Ollie accidentally called Jane and talked about whisky. Jane cancels lunch. They'll meet for a tea at 6 pm.
PREDICTION: 
Jane: Hi Ollie, are you in Warsaw?
Ollie: yes, just back!
Jane: and the 19th?
Ollie: nope, we have this party and you must be there, remember?
Jane: oh right! i lost my


Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


REFERENCE: Hilary has the keys to the apartment. Benjamin wants to get them and go take a nap. Hilary is having lunch with some French people at La Cantina. Hilary is meeting them at the entrance to the conference hall at 2 pm. Benjamin and Elliot might join them. They're meeting for the drinks in the evening.
PREDICTION: 
Hilary: I've got the keys. Whoever wants them can meet me at lunchtime or after
Elliot: I'm ok. We're meeting for the drinks in the evening anyway and I guess we'll be going back to the apartment together?
Hilary: Yeah,


Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


REFERENCE: Payton provides Max with websites selling clothes. Payton likes browsing and trying on the clothes but not necessarily buying them. Payton usually buys clothes and books as he loves reading.
PREDICTION: 
Max and Payton are best friends. They are both in high school and they are both in the same class. They are both in the same group of friends. They are both in the same class. They are both in the same group of friends. They are both in the same group of friends


Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


REFERENCE: Rita and Tina are bored at work and have still 4 hours left.
PREDICTION: 
Rita and Tina are two of the most popular girls in school. They are both very good at their respective subjects and are very popular with the boys. They are also very good friends.
Rita: I'm so bloody tired. Falling asleep at work. :-(
Tina:


Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


REFERENCE: Beatrice wants to buy Leo a scarf, but he doesn't like scarves. She cares about his health and will buy him a scarf no matter his opinion.
PREDICTION: 
Leo and Beatrice are in a shop, and Beatrice is trying to convince Leo to buy a scarf.
Notes:
- This is a one-shot.
- I'm not sure if I should have written this as a one-shot or a sequel to "The Scar


Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


REFERENCE: Eric doesn't know if his parents let him go to Ivan's brother's wedding. Ivan will talk to them.
PREDICTION: 
Eric and Ivan are best friends.
Eric: hey eric
Ivan: yea
Ivan: so youre coming to the wedding
Eric: your brother's
Ivan: yea
Ivan: i dont know mannn
Ivan: YOU DONT KNOW??



Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


REFERENCE: Wanda wants to throw a party. She asks Gina to borrow her father's car and go do groceries together. They set the date for Friday. 
PREDICTION: 
Gina and Wanda are best friends. They are both in love with the same guy, but they are not allowed to be together. They are both in love with the same guy, but they are not allowed to be together. They are both in love with the same guy, but they


Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


REFERENCE: Martin wrote a short review and won 2 cinema tickets on FB. Martin wants Aggie to go with him this week for the new film with Redford.
PREDICTION: 
Aggie and Martin are in a cinema watching a movie.
Aggie: so what did you do
Martin: just write a short review and that's it
Aggie: well done :) so what and when. and where?
Martin: the new film with Redford
A


Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


REFERENCE: Charlee is attending Portuguese theater as a subject at university. He and other students are preparing a play by Mrożek translated into Portuguese.
PREDICTION: 
Curtis: I'm in class. Theatre in Portuguese lol
Charlee: Yes. One of my subjects at the university that I attend is portuguese theatre. We are preparing a performance
Curtis: What performance is this? Are you devising it?
Charlee


Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


REFERENCE: Ella rented a car, this makes things much faster for her and Tom. 
PREDICTION: 
Ella is driving to the airport. She is going to fly to New York. She is going to meet Tom.
Ella: I’m going to New York.
Tom: I’m going to New York.
Ella: I’m going to


Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


REFERENCE: Paul is going to share his Netflix account with Luke. In exchange Luke is going to contribute to the subscription. Paul will send Luke his bank details. Paul is on vacation with his girlfriend till tomorrow.
PREDICTION: 
Paul: I am the one!
Luke: yes, I am the one!
Paul: sure, I will send you the login and password on sunday
Luke: ok we can talk tomorrow
Paul: i don't really remember it now
Luke: send me also the bank


Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


REFERENCE: Greg and Betsy have a lot of work today, so they cannot pick up Johnny from the kindergarten. However, it's Greg's turn to do it. Greg will try to find a solution.
PREDICTION: 
Betsy: I'm sorry, Greg. I can't. I'm working long hours.
Greg: I'm sorry, honey. I'll see what I can do.
Betsy: What if I can't?
Greg: Can't you, really?
Betsy:


Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


REFERENCE: Ethan, Toby and Marshall are making fun of Scott.
PREDICTION: 
Ethan and Marshall are at a bar, and Ethan is trying to get Marshall to go to a party with him.
Notes:
I'm not sure if I'm going to continue this, but I'm going to try.
(See the end of the chapter for more notes.)


Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


REFERENCE: Igor has a lot of work on his notice period and he feels demotivated. John thinks he should do what he has to do nevertheless. 
PREDICTION: 
Igor: I'm so tired of this. 
John: I know, I know. 
Igor: I'm so tired of this. 
John: I know, I know. 
Igor: I'm so tired of this. 
John: I


Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


REFERENCE: Clara is rewatching Dear White People and strongly recommends it to Neela.
PREDICTION: 
Neela: I'm not sure if I should be happy or sad that I've been watching Dear White People on Netflix.
Clara: I think you should be happy, it's a good show!
Neela: I'm not sure if I should be happy or sad that I
REFERENCE: Mike took his car into garage today. Ernest is relieved as someone had just crashed into a red Honda which looks like Mike's. 
PREDICTION: 
Mike is a car enthusiast. He has a red Honda Civic. He is a good friend of Ernest’s. Ernest is a car enthusiast. He has a red Honda Civic. He is a good friend of Mike’s.
Ernest: hey Mike, did you park

ROUGE (Base Model): {'rouge1': np.float64(0.22137185667602288), 'rouge2': np.float64(0.05648804978313898), 'rougeL': np.float64(0.17140968762778047), 'rougeLsum': np.float64(0.18029387498729926)}


# Preprocessing & Tokenizing the data

In [4]:
# ========================
# STEP 4: Preprocess Function
# ========================

def preprocess(examples):
    prompts = [f"Summarize this dialogue:\n\n{d}\n\nSummary:" for d in examples["dialogue"]]
    tokenizer.pad_token = tokenizer.eos_token
    inputs = tokenizer(prompts, truncation=True, padding="max_length", max_length=512)
    targets = tokenizer(examples["summary"], truncation=True, padding="max_length", max_length=60)
    inputs["labels"] = targets["input_ids"]
    return inputs

train_data = dataset["train"].map(preprocess, batched=True)


Map:   0%|          | 0/14731 [00:00<?, ? examples/s]

#LoRA Fine Tuning

A data collator takes tokenized examples from your dataset and turns them into mini-batches that the model can consume. it handles,

1. Padding sequences in the batch so they’re the same length.

2. Shifting tokens to create inputs and labels for language modeling.

3. Optionally applying masking(MLM)

mlm=True → Masked Language Modeling (MLM), like BERT training.

Randomly mask tokens (e.g., turn “Paris” → “[MASK]”) and ask model to predict them using the available sentence. (BERT MLM - Guessing the masking word)

mlm=False → Causal Language Modeling (CLM), like GPT/Falcon training. (Here we're predicting the next word)

Model predicts the next token in sequence, without random masking.

Since Falcon is a causal LM, we set mlm=False.

In [5]:

# ========================
# STEP 5: LoRA Fine-tuning
# ========================

from peft import LoraConfig, get_peft_model
from transformers import TrainingArguments, Trainer, DataCollatorForLanguageModeling

lora_model = AutoModelForCausalLM.from_pretrained("tiiuae/falcon-rw-1b").to("cuda")  # TinyLlama/TinyLlama-1.1B-Chat-v1.0
lora_model.gradient_checkpointing_enable()

lora_config = LoraConfig(
    r=8,
    lora_alpha=16, # a scaling factor that controls how much the LoRA update influences the original weight. High alpha have a strong influence where as lower alpha will have lesser influence.
    target_modules=["query_key_value"],
    lora_dropout=0.1,
    task_type="CAUSAL_LM"
)

lora_model = get_peft_model(lora_model, lora_config)

collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm=False)

training_args = TrainingArguments(
    output_dir="./lora-out",
    num_train_epochs=1,
    per_device_train_batch_size=8,  # each GPU trains with 4 examples at once, # Double the batch size to saturate GPU utility
    save_steps=1000,    # Saving is slow; do it less often
    logging_steps=10,  # Frequent logging helps you see progress sooner
    fp16=True,       # Use half-precision floating point (16 bits instead of 32). Saves GPU memory and speeds things up
    save_total_limit=1,
    report_to="none"
)

trainer = Trainer(model=lora_model, args=training_args, train_dataset=train_data, data_collator=collator)
trainer.train()


Loading weights:   0%|          | 0/292 [00:00<?, ?it/s]

The tied weights mapping and config for this model specifies to tie transformer.word_embeddings.weight to lm_head.weight, but both are present in the checkpoints, so we will NOT tie them. You should update the config with `tie_word_embeddings=False` to silence this warning


Step,Training Loss
10,2.703729
20,2.551997
30,2.406206
40,2.356559
50,2.375824
60,2.350128
70,2.282506
80,2.275391
90,2.268057
100,2.250728


TrainOutput(global_step=1842, training_loss=2.2206120133788265, metrics={'train_runtime': 4715.6357, 'train_samples_per_second': 3.124, 'train_steps_per_second': 0.391, 'total_flos': 5.942698265542656e+16, 'train_loss': 2.2206120133788265, 'epoch': 1.0})

In [30]:
lora_score = evaluate_model(lora_model, dataset["test"], tokenizer)
print("\n" + "="*30)
print("ROUGE (LoRAl):", lora_score)

REFERENCE: Hannah needs Betty's number but Amanda doesn't have it. She needs to contact Larry.
PREDICTION: 
If you can be the best place that he's the following.
-6.
- 
- 1- freely for the main difference.
- ery.There is the same day of the main purpose of my sister and I“there are also a new look at the process.
I have to the most important to be all of the past few days ago.
Andrea and her father.
- 
I’t even though he”


REFERENCE: Eric and Rob are going to watch a stand-up on youtube.
PREDICTION: 
Famous.
A.
-style.
- --------------
Podcasts.
- х�
-     
-  
- 
- Research Centre for your own life and I love this time.
The first on the most important to the area, and they have been to the world. I was it’t want to the idea of your property.
-6-
-8.
the- Research,



ROUGE (LoRAl): {'rouge1': np.float64(0.11715828832571665), 'rouge2': np.float64(0.0), 'rougeL': np.float64(0.07062733693394266), 'rougeLsum': np.float64(0.09991690901537183)}


#QLoRA - Fine Tuning
(Run if you want the Quantized model)

In [ ]:
# ========================
# STEP 6: QLoRA Fine-tuning
# ========================

'''
Below QLoRA function prepares a model that was loaded in 4-bit or 8-bit (quantized).

Most weights are in int4 (compressed to save GPU memory).

But some parts (like layer norms, output heads) still need to be in float32 or float16 for training stability.
'''

from transformers import BitsAndBytesConfig
from peft import prepare_model_for_kbit_training

bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_use_double_quant=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.bfloat16
)

qlora_model = AutoModelForCausalLM.from_pretrained("tiiuae/falcon-rw-1b", quantization_config=bnb_config, device_map="auto")
qlora_model = prepare_model_for_kbit_training(qlora_model)
qlora_model = get_peft_model(qlora_model, lora_config)

trainer = Trainer(model=qlora_model, args=training_args, train_dataset=train_data, data_collator=collator)
trainer.train()

qlora_score = evaluate_model(qlora_model, dataset["test"], tokenizer)
print("ROUGE (QLoRA):", qlora_score)

Push the LoRA Adapters to HuggingFace Hub

In [34]:
!huggingface-cli login

lora_model.push_to_hub("jaich/falcon_1b_LoRA_Adapter_fp16_bit_Summarizer")

tokenizer.push_to_hub("jaich/falcon_1b_LoRA_Adapter_fp16_bit_Summarizer")



Hint: `hf` is already installed! Use it directly.

Hint: Examples:
  hf auth login
  hf download unsloth/gemma-4-31B-it-GGUF
  hf upload my-cool-model . .
  hf models ls --search "gemma"
  hf repos ls --format json
  hf jobs run python:3.12 python -c 'print("Hello!")'
  hf --help



HfHubHTTPError: Client error '401 Unauthorized' for url 'https://huggingface.co/api/repos/create' (Request ID: Root=1-69e22714-13fd3dc32c8bcd587daef532;6f8b856c-4901-4cc9-9ba1-7de27bff719e)
For more information check: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/401

Invalid username or password.

# What happens when someone loads it back?

1. Load the original base model (1B params).

2. Load your LoRA adapter weights (MBs). (adapter_config.json) ()

3. Combine them at runtime.

So users still need access to the base model checkpoint.

In [ ]:
from peft import PeftModel
from transformers import AutoModelForCausalLM

base_model = AutoModelForCausalLM.from_pretrained("tiiuae/falcon-rw-1b")
model = PeftModel.from_pretrained(base_model, "jaich/falcon_1b_LoRA_Adapter_fp16_bit_Summarizer")

Merge the base model with LoRA updated weight during the training itself

In [ ]:
new_model = "falcon-1B-merged-weight-finetune-LoRA-adapter-weights"

trainer.model.save_pretrained(new_model)

from peft import LoraConfig, PeftModel
# Reload model in FP16 and merge it with LoRA weights
base_model = AutoModelForCausalLM.from_pretrained(
    "tiiuae/falcon-rw-1b",
    low_cpu_mem_usage=True,
    return_dict=True,
    torch_dtype=torch.float16,
    device_map="auto",
)
model = PeftModel.from_pretrained(base_model, new_model)
model = model.merge_and_unload()

# Reload tokenizer to save it
tokenizer = AutoTokenizer.from_pretrained("tiiuae/falcon-rw-1b", trust_remote_code=True)
tokenizer.pad_token = tokenizer.eos_token
tokenizer.padding_side = "right"

#Understand the parameters changes using LoRA adapters

In [22]:
# =====================
# Check Trainable Parameters
# =====================
def print_trainable_parameters(model):
    trainable = 0
    all_params = 0
    for _, param in model.named_parameters():
        all_params += param.numel()
        if param.requires_grad:
            trainable += param.numel()
    print(f"Trainable params: {trainable:,}")
    print(f"All params: {all_params:,}")
    print(f"Trainable%: {100 * trainable / all_params:.4f}%")

print_trainable_parameters(lora_model)


Trainable params: 1,572,864
All params: 1,416,220,672
Trainable%: 0.1111%


**Understand the results:**

All params: 1,416,220,672 → total Falcon weights (frozen + trainable).

Trainable params: 1,572,864 → LoRA adapter weights only.

Trainable %: 0.11118% → you’re updating ~1 in 900 weights (about 0.11%) instead of full fine-tuning.


“We fine-tune just 1.57M / 1.4B ≈ 0.11% of the model — roughly 1 out of every 900 weights